File: conformance/python/events.py
"""Conformance consumer: events sample, Python target.

Exercises the ABI 2 events surface through the generated ctypes wrapper: the
`Subscriber` callback interface implemented by subclassing the generated ABC
(every method is called with the right arguments, the `EventBus` handed to
`on_attached` is a usable wrapper that owns its own strong reference), the
`Delivery` return steering `publish`'s accepted count, the reference-counted
`EventBus` object (`with` statement, idempotent `close()`, use after close
rejected), the asyncio-bridged `publish_later` method, the `messages()`
iterator, the `last_message()` optional record, `route_once` releasing a
subscriber it doesn't retain, and a subscriber that raises a Python exception
surfacing to the caller as `WeaveFFIError` with `FOREIGN_ERROR_CODE` (-4)
without taking the interpreter down. The generated package is placed on
sys.path via WV_PY; the cdylib is selected with WEAVEFFI_LIBRARY.
"""
import asyncio
import gc
import os
import sys
import weakref

sys.path.insert(0, os.environ["WV_PY"])

import events as wv  # noqa: E402


def check(cond: bool, what: str) -> None:
    if not cond:
        print(f"python/events: FAIL: {what}", file=sys.stderr)
        sys.exit(1)


class Recorder(wv.Subscriber):
    """A subscriber that records every callback it receives.

    `skip_topic` is routed Skip, `stop_topic` AcceptAndStop, `fail_topic`
    raises from `route`, `fail_text` raises from `on_message`, and
    everything else is Accept. The bus received in `on_attached` is kept so
    the consumer can prove it stays alive independently of the caller's
    wrapper.
    """

    def __init__(self, name: str, skip_topic: str = "", stop_topic: str = "",
                 fail_topic: str = "", fail_text: str = "") -> None:
        self.name = name
        self.skip_topic = skip_topic
        self.stop_topic = stop_topic
        self.fail_topic = fail_topic
        self.fail_text = fail_text
        self.routed: list[str] = []
        self.messages: list[wv.Message] = []
        self.attached: list[wv.EventBus] = []
        self.attach_counts: list[int] = []

    def route(self, topic: str) -> wv.Delivery:
        self.routed.append(topic)
        if topic == self.fail_topic:
            raise ValueError(f"{self.name} rejected topic {topic}")
        if topic == self.skip_topic:
            return wv.Delivery.Skip
        if topic == self.stop_topic:
            return wv.Delivery.AcceptAndStop
        return wv.Delivery.Accept

    def on_message(self, message: wv.Message) -> int:
        if message.text == self.fail_text:
            raise RuntimeError(f"{self.name} choked on {message.text}")
        self.messages.append(message)
        return len(self.messages)

    def on_attached(self, bus: wv.EventBus) -> None:
        # The wrapper is usable right away: subscribe has not pushed us yet.
        self.attach_counts.append(bus.subscriber_count())
        self.attached.append(bus)


class Exploding(wv.Subscriber):
    """A subscriber whose `on_attached` raises, so `subscribe` itself fails."""

    def route(self, topic: str) -> wv.Delivery:
        return wv.Delivery.Accept

    def on_message(self, message: wv.Message) -> int:
        return 0

    def on_attached(self, bus: wv.EventBus) -> None:
        bus.close()
        raise KeyError("refusing to attach")


def main() -> None:
    check(wv.Delivery.Accept == 0 and wv.Delivery.Skip == 1
          and wv.Delivery.AcceptAndStop == 2, "Delivery discriminants")

    with wv.EventBus() as bus:
        check(bus.subscriber_count() == 0, "fresh bus has no subscribers")
        check(bus.last_message() is None, "fresh bus has no last message")
        check(list(bus.messages()) == [], "fresh bus has no messages")

        quiet = Recorder("quiet", skip_topic="quiet")
        loud = Recorder("loud", stop_topic="stop")
        tail = Recorder("tail")

        check(bus.subscribe(quiet) == 1, "first subscribe returns 1")
        check(bus.subscribe(loud) == 2, "second subscribe returns 2")
        check(bus.subscribe(tail) == 3, "third subscribe returns 3")
        check(bus.subscriber_count() == 3, "subscriber_count after three subscribes")

        # on_attached fired once per subscriber, before the push, with a
        # live bus wrapper.
        check(quiet.attach_counts == [0], f"quiet attach count {quiet.attach_counts}")
        check(loud.attach_counts == [1], f"loud attach count {loud.attach_counts}")
        check(tail.attach_counts == [2], f"tail attach count {tail.attach_counts}")
        check(len(quiet.attached) == 1, "quiet kept its attached bus")
        check(isinstance(quiet.attached[0], wv.EventBus),
              "on_attached receives an EventBus wrapper")
        # The adopted reference is independent of the caller's wrapper: it is
        # a different Python object on the same underlying bus.
        check(quiet.attached[0] is not bus, "attached wrapper is a distinct object")
        check(quiet.attached[0].subscriber_count() == 3,
              "attached bus wrapper observes the shared state")

        # Accept everywhere: every subscriber gets it.
        n = bus.publish("news", "hello", ["a", "b"])
        check(n == 3, f"publish news accepted {n}")
        check(quiet.routed == ["news"] and loud.routed == ["news"]
              and tail.routed == ["news"], "route called on each subscriber")
        for sub in (quiet, loud, tail):
            check(len(sub.messages) == 1, f"{sub.name} received one message")
            m = sub.messages[0]
            check(m == wv.Message(seq=1, topic="news", text="hello", tags=["a", "b"]),
                  f"{sub.name} message {m}")

        # Skip steers the count: quiet drops its own topic.
        n = bus.publish("quiet", "psst", [])
        check(n == 2, f"publish quiet accepted {n}")
        check(len(quiet.messages) == 1, "quiet skipped its topic")
        check(loud.messages[-1].seq == 2 and loud.messages[-1].tags == [],
              "seq advances and empty tags round-trip")

        # AcceptAndStop: loud accepts and stops delivery to tail.
        n = bus.publish("stop", "last", ["z"])
        check(n == 2, f"publish stop accepted {n}")
        check(len(quiet.messages) == 2, "quiet accepted stop")
        check(len(loud.messages) == 3, "loud accepted stop")
        check(len(tail.messages) == 2, "tail was not reached after AcceptAndStop")
        check(tail.routed == ["news", "quiet"], f"tail routed {tail.routed}")

        # Async publish: resolved from the producer's thread through asyncio.
        n = asyncio.run(bus.publish_later("later", "async"))
        check(n == 3, f"publish_later accepted {n}")
        check(tail.messages[-1] == wv.Message(4, "later", "async", []),
              f"async message {tail.messages[-1]}")

        # Iterator and optional record.
        check(list(bus.messages()) == ["hello", "psst", "last", "async"],
              "messages() streams every text in order")
        it = bus.messages()
        check(next(it) == "hello", "manual next on iterator")
        it.close()
        it.close()
        last = bus.last_message()
        check(last == wv.Message(seq=4, topic="later", text="async", tags=[]),
              f"last_message {last}")

        # A subscriber that raises from route aborts publish with the foreign
        # error code, and the bus is still usable afterwards.
        grumpy = Recorder("grumpy", fail_topic="boom", fail_text="gag")
        check(bus.subscribe(grumpy) == 4, "grumpy subscribed")
        try:
            bus.publish("boom", "x", [])
            check(False, "expected WeaveFFIError from a raising route")
        except wv.WeaveFFIError as exc:
            check(exc.code == wv.WeaveFFIError.FOREIGN_ERROR_CODE == -4,
                  f"foreign error code {exc.code}")
            check("grumpy rejected topic boom" in exc.message,
                  f"foreign error message {exc.message!r}")
        # Earlier subscribers were already delivered before grumpy raised.
        check(len(quiet.messages) == 4, "quiet got the aborted message first")
        # A raise from on_message is reported the same way.
        try:
            bus.publish("ok", "gag", [])
            check(False, "expected WeaveFFIError from a raising on_message")
        except wv.WeaveFFIError as exc:
            check(exc.code == -4, f"on_message foreign code {exc.code}")
            check("choked on gag" in exc.message, f"on_message message {exc.message!r}")
        # The log still recorded both aborted publishes.
        check(bus.last_message().text == "gag" and bus.last_message().seq == 6,
              "aborted publishes are still logged")
        n = bus.publish("ok", "recovered", [])
        check(n == 4, f"bus usable after foreign error, accepted {n}")
        check(grumpy.messages[-1].text == "recovered", "grumpy delivered after recovering")

        # A raise from on_attached fails subscribe itself and leaves the
        # subscriber unretained (its free runs during the abort).
        exploding = Exploding()
        exploding_ref = weakref.ref(exploding)
        try:
            bus.subscribe(exploding)
            check(False, "expected WeaveFFIError from a raising on_attached")
        except wv.WeaveFFIError as exc:
            check(exc.code == -4, f"on_attached foreign code {exc.code}")
            check("refusing to attach" in exc.message, f"on_attached message {exc.message!r}")
        check(bus.subscriber_count() == 4, "failed subscribe did not retain")
        del exploding
        gc.collect()
        check(exploding_ref() is None, "unretained subscriber released after failed subscribe")

        # route_once does not retain the subscriber: its free runs before the
        # call returns, so nothing but our local keeps it alive.
        probe = Recorder("probe", skip_topic="quiet")
        probe_ref = weakref.ref(probe)
        check(wv.route_once(probe, "quiet") == wv.Delivery.Skip, "route_once Skip")
        check(wv.route_once(probe, "other") == wv.Delivery.Accept, "route_once Accept")
        check(probe.routed == ["quiet", "other"], f"route_once routed {probe.routed}")
        check(probe.attached == [], "route_once never attaches")
        del probe
        gc.collect()
        check(probe_ref() is None, "route_once released its subscriber")
        try:
            wv.route_once(Recorder("x", fail_topic="bad"), "bad")
            check(False, "expected WeaveFFIError from route_once")
        except wv.WeaveFFIError as exc:
            check(exc.code == -4, f"route_once foreign code {exc.code}")

        # Clearing the subscribers runs each free: once the consumer drops its
        # own references, nothing keeps the implementations alive.
        refs = [weakref.ref(s) for s in (quiet, loud, tail, grumpy)]
        kept_bus = quiet.attached[0]
        del quiet, loud, tail, grumpy, sub
        gc.collect()
        check(all(r() is not None for r in refs), "retained subscribers stay alive")
        bus.clear_subscribers()
        check(bus.subscriber_count() == 0, "clear_subscribers empties the bus")
        gc.collect()
        check(all(r() is None for r in refs), "cleared subscribers were freed")

        # The bus reference adopted through on_attached outlives the caller's
        # wrapper: closing ours leaves the object alive through the other.
        bus.close()
        bus.close()
        try:
            bus.subscriber_count()
            check(False, "expected use-after-close error")
        except wv.WeaveFFIError as exc:
            check("after close" in exc.message, f"use-after-close message {exc.message!r}")
        check(list(kept_bus.messages()) == ["hello", "psst", "last", "async", "x", "gag", "recovered"],
              "bus alive through the callback-adopted reference")
        kept_bus.close()
        kept_bus.close()
    # Leaving the with block closes an already-closed bus: a no-op.

    # A fresh bus is independent state.
    other = wv.EventBus()
    check(other.subscriber_count() == 0 and other.last_message() is None,
          "second bus is independent")
    other.close()

    print("python/events: OK")


main()

File: conformance/node/events.js
// Conformance consumer: events sample, Node (N-API) target.
//
// Drives the ABI 2 callback-interface surface through the generated wrapper
// layer (index.js): a `Subscriber` implemented as a plain JS object (route ->
// Delivery, onMessage -> running count, onAttached receiving the EventBus
// object the producer hands over), the reference-counted `EventBus` class
// (constructor, methods, idempotent `close()`, use-after-close trap), the
// `Delivery` return steering `publish`'s accepted count, the Promise-backed
// `publishLater` whose callbacks hop from the producer thread onto the JS
// thread, the lazy iterable `messages()`, the nullable `lastMessage()`
// record, the free function `routeOnce`, and a subscriber that throws: the
// exception surfaces to the caller as a WeaveFFIError with code -4 (foreign
// error) without aborting the process. i64 values cross as BigInt. The
// harness passes the built addon via WV_ADDON; the generated loader honors
// WEAVEFFI_ADDON.

'use strict';

const path = require('path');

const ADDON = path.resolve(process.env.WV_ADDON);
process.env.WEAVEFFI_ADDON = ADDON;
// WV_ADDON = <target>/conformance-build/node-events/build/Release/index.node;
// the generated files sit at <target>/conformance-gen/events/node/.
const wv = require(
  path.resolve(ADDON, '../../../../..', 'conformance-gen', 'events', 'node', 'index.js')
);

let failures = 0;
function expect(cond, msg) {
  if (!cond) {
    console.error('assertion failed: ' + msg);
    failures++;
  }
}

// The C-style enum is exported as a frozen object with forward and reverse
// mappings, the runtime value `types.d.ts` declares as `export enum`.
const Delivery = wv.Delivery;
expect(Delivery.Accept === 0 && Delivery.Skip === 1 && Delivery.AcceptAndStop === 2, 'Delivery values');
expect(Delivery[2] === 'AcceptAndStop', 'Delivery reverse mapping');
expect(Object.isFrozen(Delivery), 'Delivery is frozen');

// A subscriber as a class: `route` steers delivery per topic, `onMessage`
// records what it saw and returns its running count (an i64, so a BigInt),
// and `onAttached` adopts the bus object the producer hands over.
class RecordingSubscriber {
  constructor(name, opts) {
    this.name = name;
    this.skipTopic = (opts && opts.skip) || null;
    this.stopTopic = (opts && opts.stop) || null;
    this.throwTopic = (opts && opts.throwOn) || null;
    this.routed = [];
    this.received = [];
    this.attachedBuses = [];
    this.attachedCounts = [];
  }
  route(topic) {
    this.routed.push(topic);
    if (topic === this.throwTopic) {
      throw new Error('subscriber ' + this.name + ' rejects ' + topic);
    }
    if (topic === this.skipTopic) return Delivery.Skip;
    if (topic === this.stopTopic) return Delivery.AcceptAndStop;
    return Delivery.Accept;
  }
  onMessage(message) {
    this.received.push(message);
    return BigInt(this.received.length);
  }
  onAttached(bus) {
    // The object is usable right away: `subscribe` attaches before it
    // appends, so the count seen here is the pre-subscribe count.
    expect(bus instanceof wv.EventBus, this.name + ': onAttached receives an EventBus');
    this.attachedCounts.push(bus.subscriberCount());
    this.attachedBuses.push(bus);
  }
}

(async () => {
  const bus = new wv.EventBus();
  expect(bus instanceof wv.EventBus, 'constructor returns an EventBus');
  expect(bus.subscriberCount() === 0n, 'fresh bus has no subscribers');
  expect(bus.lastMessage() === null, 'lastMessage is null before any publish');
  expect([...bus.messages()].length === 0, 'messages() empty before any publish');

  // subscribe() returns the new subscriber count and calls onAttached with
  // the bus object (one strong reference transferred to the consumer).
  const a = new RecordingSubscriber('a', { skip: 'quiet', stop: 'stop' });
  const b = new RecordingSubscriber('b');
  expect(bus.subscribe(a) === 1n, 'subscribe a returns 1n');
  expect(bus.subscribe(b) === 2n, 'subscribe b returns 2n');
  expect(bus.subscriberCount() === 2n, 'subscriberCount == 2n');
  expect(a.attachedBuses.length === 1, 'a attached once');
  expect(a.attachedCounts[0] === 0n, `a saw 0 subscribers at attach (got ${a.attachedCounts[0]})`);
  expect(b.attachedCounts[0] === 1n, `b saw 1 subscriber at attach (got ${b.attachedCounts[0]})`);

  // Both accept: two deliveries, one Message each with every field decoded.
  const n1 = bus.publish('news', 'hello', ['x', 'y']);
  expect(typeof n1 === 'bigint', 'publish returns a BigInt');
  expect(n1 === 2n, `publish(news) accepted by 2 (got ${n1})`);
  expect(a.received.length === 1 && b.received.length === 1, 'both received news');
  const m = a.received[0];
  expect(m.seq === 1n, `message seq is 1n (got ${m.seq})`);
  expect(m.topic === 'news', 'message topic');
  expect(m.text === 'hello', 'message text');
  expect(
    Array.isArray(m.tags) && m.tags.length === 2 && m.tags[0] === 'x' && m.tags[1] === 'y',
    `message tags (got ${JSON.stringify(m.tags)})`
  );
  expect(a.routed.length === 1 && a.routed[0] === 'news', 'route called with topic');

  // Skip steers the count: a skips "quiet", only b accepts.
  const n2 = bus.publish('quiet', 'psst', []);
  expect(n2 === 1n, `publish(quiet) accepted by 1 (got ${n2})`);
  expect(a.received.length === 1, 'a skipped quiet');
  expect(b.received.length === 2 && b.received[1].seq === 2n, 'b received quiet with seq 2n');
  expect(b.received[1].tags.length === 0, 'empty tags list decodes as []');

  // AcceptAndStop: a accepts and the bus stops before reaching b.
  const n3 = bus.publish('stop', 'last', ['z']);
  expect(n3 === 1n, `publish(stop) accepted by 1 (got ${n3})`);
  expect(a.received.length === 2 && a.received[1].text === 'last', 'a received stop');
  expect(b.received.length === 2, 'b not reached after AcceptAndStop');
  expect(b.routed.length === 2, 'b.route not consulted after stop');

  // The adopted bus is the same underlying object: it observes the same log.
  const viaA = a.attachedBuses[0];
  expect(viaA.subscriberCount() === 2n, 'adopted bus sees both subscribers');
  const viaAMsgs = [...viaA.messages()];
  expect(
    viaAMsgs.join('|') === 'hello|psst|last',
    `adopted bus shares the log (got ${JSON.stringify(viaAMsgs)})`
  );
  const lastViaA = viaA.lastMessage();
  expect(lastViaA !== null && lastViaA.seq === 3n && lastViaA.text === 'last', 'adopted bus lastMessage');
  // Releasing the adopted references leaves the original alive.
  for (const adopted of a.attachedBuses.concat(b.attachedBuses)) {
    adopted.close();
    adopted.close();
  }
  expect(bus.subscriberCount() === 2n, 'bus alive after adopted wrappers closed');

  // Async publish: the producer thread calls route/onMessage, which hop to
  // the JS thread while this frame awaits.
  const later = bus.publishLater('news', 'later');
  expect(later instanceof Promise, 'publishLater returns a Promise');
  const n4 = await later;
  expect(n4 === 2n, `publishLater accepted by 2 (got ${n4})`);
  expect(a.received.length === 3 && a.received[2].text === 'later', 'a received the async message');
  expect(b.received.length === 3 && b.received[2].seq === 4n, 'b received the async message');
  expect(a.received[2].tags.length === 0, 'publishLater sends no tags');

  // Iterator and nullable record.
  const msgs = [...bus.messages()];
  expect(
    msgs.join('|') === 'hello|psst|last|later',
    `messages() in order (got ${JSON.stringify(msgs)})`
  );
  const last = bus.lastMessage();
  expect(last !== null && typeof last === 'object', 'lastMessage present');
  expect(last.seq === 4n && last.topic === 'news' && last.text === 'later', 'lastMessage fields');
  expect(Array.isArray(last.tags) && last.tags.length === 0, 'lastMessage tags empty');

  // Early return closes the underlying producer iterator without draining.
  const it = bus.messages()[Symbol.iterator]();
  const first = it.next();
  expect(!first.done && first.value === 'hello', 'lazy first element');
  expect(typeof it.return === 'function', 'iterator has return()');
  it.return();
  expect(it.next().done === true, 'iterator done after return()');

  // A subscriber whose route throws: the producer aborts publish and the
  // caller sees the foreign error (code -4) carrying the thrown message.
  const bad = new RecordingSubscriber('bad', { throwOn: 'boom' });
  expect(bus.subscribe(bad) === 3n, 'subscribe bad returns 3n');
  try {
    bus.publish('boom', 'x', []);
    expect(false, 'expected publish to throw when a subscriber throws');
  } catch (e) {
    expect(e instanceof wv.WeaveFFIError, `foreign error is a WeaveFFIError (got ${e && e.constructor.name})`);
    expect(e.code === -4, `foreign error code is -4 (got ${e.code})`);
    expect(
      typeof e.errorMessage === 'string' && e.errorMessage.includes('rejects boom'),
      `foreign error carries the JS message (got ${JSON.stringify(e.errorMessage)})`
    );
  }
  // a and b ran before bad and still accepted; the bus is intact.
  expect(a.received.length === 4 && b.received.length === 4, 'earlier subscribers still delivered');
  expect(bus.subscriberCount() === 3n, 'bus intact after foreign error');
  expect(bus.publish('ok', 'y', []) === 3n, 'publish works again after foreign error');

  // The same through the async path: the Promise rejects with code -4.
  try {
    await bus.publishLater('boom', 'z');
    expect(false, 'expected publishLater to reject when a subscriber throws');
  } catch (e) {
    expect(e instanceof wv.WeaveFFIError && e.code === -4, `async foreign error code -4 (got ${e && e.code})`);
    expect(e.errorMessage.includes('rejects boom'), 'async foreign error carries the JS message');
  }

  // A wrong return type from the callback is also a foreign error.
  const weird = { route() { return 'not a number'; }, onMessage() { return 0n; }, onAttached() {} };
  try {
    wv.routeOnce(weird, 'any');
    expect(false, 'expected routeOnce to throw for a bad return type');
  } catch (e) {
    expect(e instanceof wv.WeaveFFIError && e.code === -4, `bad return type is code -4 (got ${e && e.code})`);
  }

  // routeOnce: a free function taking the callback interface; no bus involved.
  const solo = new RecordingSubscriber('solo', { skip: 'quiet' });
  expect(wv.routeOnce(solo, 'quiet') === Delivery.Skip, 'routeOnce(quiet) == Skip');
  expect(wv.routeOnce(solo, 'news') === Delivery.Accept, 'routeOnce(news) == Accept');
  expect(solo.routed.length === 2 && solo.attachedBuses.length === 0, 'routeOnce calls route only');
  // A plain object literal works as an implementation too.
  const literal = { route: () => Delivery.AcceptAndStop, onMessage: () => 1n, onAttached: () => {} };
  expect(wv.routeOnce(literal, 'x') === Delivery.AcceptAndStop, 'object literal subscriber');

  // clearSubscribers drops every retained subscriber.
  bus.clearSubscribers();
  expect(bus.subscriberCount() === 0n, 'subscriberCount == 0n after clear');
  expect(bus.publish('news', 'nobody', []) === 0n, 'publish with no subscribers accepts 0');
  const all = [...bus.messages()];
  expect(all.length === 8, `log records every publish, aborted ones included (got ${all.length})`);

  // Idempotent close; use after close is a marshalling error (-3).
  bus.close();
  bus.close();
  try {
    bus.subscriberCount();
    expect(false, 'expected throw for use after close');
  } catch (e) {
    expect(e instanceof wv.WeaveFFIError && e.code === -3, `use after close is code -3 (got ${e && e.code})`);
  }
  if (typeof Symbol.dispose === 'symbol') {
    const disposable = new wv.EventBus();
    disposable[Symbol.dispose]();
    disposable[Symbol.dispose]();
  }

  if (failures > 0) process.exit(1);
  console.log('node/events: OK');
})().catch((e) => {
  console.error(e);
  process.exit(1);
});
